File: LeetCode 30 Day Challenge/11-diameterOfBinaryTree.js
class TreeNode {
    constructor(val) {
        this.val = val;
        this.left = this.right = null;
    }
}

const diameterOfBinaryTree = (root) => {
    let diameter = 0
    dfs(root)
    
    return diameter   
}

const dfs = (node) => {
    if (!node) return 0
    
    let left = dfs(node.left)
    let right = dfs(node.right)
    
    diameter = Math.max(left + right, diameter)
    
    return 1 + Math.max(left, right)
}

File: LeetCode 30 Day Challenge/11-diameter_of_binary_tree.py
class TreeNode(object):
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

class Solution(object):
    def diameterOfBinaryTree(self, root):
        self.diameter = 0
        
        self.dfs(root)
        
        return self.diameter
    
    def dfs(self, node):
        if not node:
            return 0
        
        left = self.dfs(node.left)
        right = self.dfs(node.right)
        
        self.diameter = max(left + right, self.diameter)
        return 1 + max(left, right)